### ISSUE TITLE
allow-clear with theme=bootstrap broken in 0.9.6

### ISSUE THREAD

**Issue by zavidovych** [angular-ui/ui-select#578]
http://plnkr.co/edit/yzoAL0PRkOyGHeuMbGj2?p=preview

![unkjfgzpb8](https://cloud.githubusercontent.com/assets/1755574/5715277/4a741a0c-9a8b-11e4-9ecb-d2a191d56c6e.png)

**Comment by artkoenig** [comment]
It does not relate to allow-clear, the bootstrap-layout is broken completely:

http://plnkr.co/edit/2mQ6S7w01mjKhWUMEf6b?p=preview

It was ok in 0.9.2

**Comment by jack-kerouac** [comment]
:+1:

**Comment by gerardp** [comment]
:+1:

**Comment by BP18** [comment]
:+1:

The [Demo Plunker](http://plnkr.co/edit/a3KlK8dKH3wwiiksDSn2?p=preview) linked to from the [README.md](https://github.com/angular-ui/ui-select/blob/0.9.6/README.md) should be updated to use the latest version of select.css and select.js. Then this issue is immediately obvious there.

1) Temp workaround for text alignment (probably a better way to fix this):
```
.ui-select-bootstrap > .ui-select-match > .ui-select-toggle {
    text-align: left;
}
```

2) Temp workaround for missing bootstrap styles (box-shadow, etc.):
Add css class `form-control` to `button` tag on line 2 of [match.tpl.html](https://github.com/angular-ui/ui-select/blob/0.9.6/src/bootstrap/match.tpl.html)

**Comment by abobwhite** [comment]
@BP18 Thanks. I just tried adding your two workaround adjustments for version 0.9.6 and it didn't appear to make any difference in the styling.

**Comment by abobwhite** [comment]
It appears that if the allowClear directive is included, there is a conditional ```ui-select-allow-clear``` CSS class added that does nothing (no usage in select.js or select.css)...

This is very frustrating.

**Comment by BP18** [comment]
Hi @abobwhite, I'm not using allow-clear, but my workarounds seem to work on the demo page.

Here's the Demo Plunker as-is other than upgrading to 0.9.6 so I can reproduce the problem:
http://plnkr.co/edit/TEt1lqLYSwHo30EQdLvr?p=preview

Here's the same Plunker but with my two fixes added:
http://plnkr.co/edit/M1A5BhAK7et987Bhgih3?p=preview

The padding is still messed up, but at least the text-align and other styles are better. Don't think this is really a proper fix, but hopefully it helps figure out what's going on.

**Comment by abobwhite** [comment]
@BP18 Thanks.

I'm confused - isn't this issue surrounding the allow-clear styling? ```allow-clear``` on both plunkr's yield the same styling issue. If you weren't talking about ```allow-clear```, what styling were you trying to fix, exactly?

**Comment by BP18** [comment]
@abobwhite, in my first demo Plunker link with just the upgrade to 0.9.6, the Bootstrap theme control is messed up. Without focus on the control, the text is center aligned and the field does not have the other Bootstrap `form-control` css styles like `box-shadow` and `transition: border-color ease-in-out`.

**Comment by artkoenig** [comment]
The template did change from

![bildschirmfoto 2015-01-22 um 21 41 53](https://cloud.githubusercontent.com/assets/1738628/5864172/16e75b0c-a280-11e4-8a14-b193cca7b206.png)

to

![bildschirmfoto 2015-01-22 um 21 42 40](https://cloud.githubusercontent.com/assets/1738628/5864179/1f9508e4-a280-11e4-9357-a71d9f92ac5f.png)

could this be the cause?

**Comment by BP18** [comment]
@artkoenig, yes I think that is where it broke. Commit [6e94894](https://github.com/angular-ui/ui-select/commit/6e9489435e4bf2e8714839cf114e507a6e90db89).

Previously the button tag had the `form-control` css class applied, but with the new template it doesn't. This caused the shadow, transition, and other styles to be lost. The text-align and padding were being controlled by overrides in the select.css file, but that broke with this template change too.

**Comment by BP18** [comment]
Slightly tweaked my original fixes and submitted a pull request.

**Comment by gzsombor** [comment]
I tried to test your changeset, but the 'allow-clear' button is still broken. Is it supposed to be fixed by your changeset ?

**Comment by abobwhite** [comment]
That was my understanding as well - it sounds like there are two issues here. Broken bootstrap styles in general as well as broken allow-clear styling for bootstrap. I am most concerned about the allow clear but I'm sure both need to be addressed.

**Comment by BP18** [comment]
Yes, I think @abobwhite is correct about there being two (related?) issues here. I'm not using allow-clear and neither does the demo page, so my fix didn't address that part.

**Comment by BP18** [comment]
I played around with allow-clear=true and I see the second problem you are mentioning (the X/clear button appearing below the field instead of integrated). I think my pull request addresses the other styling issues and doesn't make the allow-clear issue worse, so hopefully it can be part of a complete solution.

Also, there was a pull request (#486) that was supposed to add allow-clear support to the bootstrap template, but it doesn't work properly.

**Comment by starf** [comment]
Thanks for all the hard work already put in, a proper fix would be awesome.
tl;dr +1

**Comment by tarlepp** [comment]
+1 for this

**Comment by tarlepp** [comment]
Meanwhile you can use this little CSS trick:

```
.ui-select-clear {
    position: relative;
    float: right;
    top: -26px;
    right: 25px;
    border: none;
    background: none;
    margin-bottom: -26px;
}
```

**Comment by dimirc** [comment]
With https://github.com/angular-ui/ui-select/pull/606 merged, can you confirm if all works ok now?

**Comment by JeromeLam** [comment]
@dimirc That merge doesn't seem to fix the problem.  Still looks like the screen grab from the original post.

**Comment by dimirc** [comment]
@JeromeLam you are right, seams that the problem is the way allow-clear was implemented at https://github.com/angular-ui/ui-select/pull/486 

I'm checking the code and maybe reverting that PR and do a different implementation

**Comment by JeromeLam** [comment]
@dimirc Looking at #486, I see the template for the clear button uses **btn btn-default**.  Using **btn-link** will at least get rid of the border box.  It may then just be a question of styling it using part of the CSS trick from tarlepp 5 days ago.

**Comment by dimirc** [comment]
@JeromeLam based on @bampakoa [comment](https://github.com/angular-ui/ui-select/pull/486#issuecomment-67307275) I pushed PR https://github.com/angular-ui/ui-select/pull/667

Check updated [plunker](http://plnkr.co/edit/QGElZ3h8FV9g1Quf38pG?p=preview) and I think now should be ok

**Comment by JeromeLam** [comment]
@dimirc It's definitely looking better now!  I suggest we add **btn-link** to the class list for the clear link so there is no odd outline and box shadow when the control is active or in focus (at least on Chrome).

I've created PR #669 tor this and forked your [plunker](http://plnkr.co/edit/KCulIssW17K2CA8WFe4z?p=preview) with the change applied.

**Comment by gzsombor** [comment]
Have you checked in firefox ? It seems that the clear button is still below the drop down, however now, it's on the right side.

**Comment by starf** [comment]
FYI: That happens only when the input is so long it overlaps the X button.

**Comment by dimirc** [comment]
Should be solved with #667
